fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingBase.java

```diff
@@ -576,44 +537,10 @@ private CompletableFuture<Void> throwUnlessNoRecordWasScanned(boolean noRecordSc
     }
 
     @Nonnull
-    @SuppressWarnings("PMD.CloseResource")
-    private CompletableFuture<Void> setScrubberTypeOrThrow(FDBRecordStore store) {
-        // HERE: The index must be readable, checked by the caller
-        //   if scrubber had already run and still have missing ranges, do nothing
-        //   else: clear ranges and overwrite type-stamp
-        IndexBuildProto.IndexBuildIndexingStamp indexingTypeStamp = getIndexingTypeStamp(store);
-        validateOrThrowEx(indexingTypeStamp.getMethod().equals(IndexBuildProto.IndexBuildIndexingStamp.Method.SCRUB_REPAIR),
-                "Not a scrubber type-stamp");
-
-        final Index index = common.getIndex(); // Note: the scrubbers do not support multi target (yet)
-        IndexingRangeSet indexRangeSet = IndexingRangeSet.forScrubbingIndex(store, index);
-        IndexingRangeSet recordsRangeSet = IndexingRangeSet.forScrubbingRecords(store, index);
-        final CompletableFuture<Range> indexRangeFuture = indexRangeSet.firstMissingRangeAsync();
-        final CompletableFuture<Range> recordRangeFuture = recordsRangeSet.firstMissingRangeAsync();
-        return indexRangeFuture.thenCompose(indexRange -> {
-            if (indexRange == null) {
-                // Here: no un-scrubbed index range was left for this call. We will
-                // erase the 'ranges' data to allow a fresh index re-scrubbing.
-                if (LOGGER.isDebugEnabled()) {
-                    LOGGER.debug(KeyValueLogMessage.build("Reset scrubber's index range")
-                            .addKeysAndValues(common.indexLogMessageKeyValues())
-                            .toString());
-                }
-                indexRangeSet.clear();
-            }
-            return recordRangeFuture.thenAccept(recordRange -> {
-                if (recordRange == null) {
-                    // Here: no un-scrubbed records range was left for this call. We will
-                    // erase the 'ranges' data to allow a fresh records re-scrubbing.
-                    if (LOGGER.isDebugEnabled()) {
-                        LOGGER.debug(KeyValueLogMessage.build("Reset scrubber's records range")
-                                .addKeysAndValues(common.indexLogMessageKeyValues())
-                                .toString());
-                    }
-                    recordsRangeSet.clear();
-                }
-            });
-        });
+    protected CompletableFuture<Void> setScrubberTypeOrThrow(FDBRecordStore store) {
```

Comment:
Can this be removed from this level, and only be defined in IndexScrubbing?

Reply:
Looking at the usage, I think it would require a lot more untangling of the relationship between `IndexingBase`. The simplest approach I can see would be to introduce two additional abstract classes, an `IndexScrubbingBase` and `IndexBuildingBase`. If you see something trivial go for it, but otherwise, I guess it will stay as-is.